File: java/03-bootstrap/src/test/java/com/fResult/bootstrap/customers/services/BaseClass.java
package com.fResult.bootstrap.customers.services;

import static org.junit.jupiter.api.Assertions.*;

import java.util.stream.Stream;
import org.junit.jupiter.api.Test;

public abstract class BaseClass {
  public abstract CustomerService getCustomerService();

  @Test
  public void insert() {
    final var actualResult = getCustomerService().save("Wick");

    assertNotNull(actualResult);
    assertEquals(1, actualResult.size());
  }

  @Test
  public void getAllCustomers() {
    Stream.of("A", "B").forEach(getCustomerService()::save);

    assertTrue(getCustomerService().findAll().size() > 2);
  }

  @Test
  public void getById() {
    final var id = getCustomerService().save("A").iterator().next().id();

    assertEquals(id, getCustomerService().findById(id).id());
  }
}

File: kotlin/03-bootstrap/src/test/kotlin/com/fResult/bootstrap/customers/services/BaseClass.kt
package com.fResult.bootstrap.customers.services

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

abstract class BaseClass {
  abstract fun getCustomerService(): CustomerService
  private lateinit var customerService: CustomerService

  @BeforeEach
  fun setUp() {
    customerService = getCustomerService()
  }

  @Test
  open fun `should insert customer`() {
    val actualResult = customerService.save("Wick")

    Assertions.assertNotNull(actualResult)
    Assertions.assertEquals(1, actualResult.size)
  }

  @Test
  fun `should retrieve all customers`() {
    listOf("A", "B").forEach { names ->
      customerService.save(names)
    }

    Assertions.assertTrue(customerService.findAll().size > 2)
  }

  @Test
  fun `should retrieve customer by id`() {
    val id = customerService.save("A").iterator().next().id

    customerService.findById(id)?.also { Assertions.assertEquals(id, it.id) }
  }
}

